torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -154,118 +298,339 @@ def cdiv(x: int, y: int):
+                **config,
+            )
     # add the unoptimized choice to mitigate performance degradation
-    choices.append(unoptimized_choice.bind((mat1, mat2, mat3), layout))
-    return autotune_select_algorithm("b2b_gemm", choices, [mat1, mat2, mat3], layout)
+    choices.append(
+        unoptimized_choice.bind((A, B, C), layout, is_left_assoc=is_left_assoc, subgraph=subgraph)
+    )
+    # autotune
+    return autotune_select_algorithm("b2b_gemm", choices, [A, B, C], layout)
 
 
-# currently it matches ((A @ B) @ C)
-# TODO: later will change to matching (A @ B) in (epilogue2 ((epilogue1 (A @ B)) @ C)) and inspecting the graph
-# TODO: match more cases such as bmm and addmm, and (A @ (B @ C)), etc.
+# match the inner mm of a potential b2b_gemm
 @register_graph_pattern(
-    CallFunction(
-        torch.ops.aten.mm, CallFunction(torch.ops.aten.mm, Arg(), Arg()), Arg()
-    ),
-    extra_check=can_apply_b2b_gemm,
+    CallFunction(torch.ops.aten.mm, Arg(), Arg()),
     pass_dict=B2B_GEMM_PASS,
 )
 def b2b_gemm(
-    match: Match, mat1: torch.fx.Node, mat2: torch.fx.Node, mat3: torch.fx.Node
+    match: Match, mat1: torch.fx.Node, mat2: torch.fx.Node
 ) -> None:
+    # match.args: list[torch.fx.Node]
+
+    def is_pointwise_node(node: torch.fx.Node) -> bool:
+        return (
+            node.op == "call_function" and
+            isinstance(node.target, torch._ops.OpOverload) and
+            (torch.Tag.pointwise in node.target.tags)
+        )
+    
+    def is_mm(node: torch.fx.Node) -> bool:
+        return (
+            node.op == "call_function" and
+            node.target == torch.ops.aten.mm.default
+        )
+
+    # the inner MM
+    inner_mm = match.nodes[-1]
+
+    # find the (candidate) outer MM
+    # In a real (A @ f(B @ C)), every path starting from (B @ C) must reach (A @ _).
+    outer_mm = None
+    node = inner_mm
+    while len(node.users) > 0:
+        node = next(iter(node.users))
+        if is_mm(node):
+            outer_mm = node
+            break
+        elif is_pointwise_node(node):
+            continue
+        else:
+            break
+    if not outer_mm:
+        return
+
+    # find the unique input node for outer_mm representing f(B @ C) in (A @ f(B @ C))
+    # we call it the "f_node"
+    # when the pattern is simply (A @ (B @ C)), f_node is just inner_mm
+    f_node = inner_mm
+    while not (next(iter(f_node.users)) is outer_mm):
+        f_node = next(iter(f_node.users))
+
+    def all_reach_via(
+        src: torch.fx.Node,
+        dst: torch.fx.Node,
+        check_intermediate: Callable[[torch.fx.Node], bool],
+        get_next: Callable[[torch.fx.Node], Iterable[torch.fx.Node]],
+    ) -> bool:
+        '''
+        checks whether every path from src reaches dst via check_intermediate nodes
+        where the direction is specified by get_next
+        '''
+        visited: Dict[torch.fx.Node, bool] = {}
+        def dfs(node: torch.fx.Node) -> bool:
```

Comment:
nit - can you do this as a loop instead of recursion? python has a recursion limit - also as soon as we hit `False` we should break out and return, this will continue to iterate

Reply:
added early break ... also there are multiple recursive functions, not sure whether we should update all to loop (using manual stacks, possibly)